### PR TITLE
Resubmit: Support negtive position arguments

### LINKED ISSUE #57736: Negative positional arguments.
**Use case**

`ORDER BY -1` means order by the last column in the SELECT list; negative numbers count from the end of the SELECT list.

### PR DESCRIPTION
**Pull request by ucasfl**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support negative positional arguments. Closes https://github.com/ClickHouse/ClickHouse/issues/57736.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 32ff152f2d7e4798a7bbc916808cc9ca883cf13e with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58292/32ff152f2d7e4798a7bbc916808cc9ca883cf13e/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58292/32ff152f2d7e4798a7bbc916808cc9ca883cf13e/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by Avogar — APPROVED** [review]

**Review by novikd — DISMISSED** [review]
This PR lacks implementation in `QueryAnalysisPass.cpp`.

**Comment by ucasfl** [comment]
> This PR lacks implementation in `QueryAnalysisPass.cpp`.

The implementation be reverted accidentally.

File: src/Analyzer/Passes/QueryAnalysisPass.cpp [hunk 1]
```diff
@@ -2156,19 +2156,31 @@ void QueryAnalyzer::replaceNodesWithPositionalArguments(QueryTreeNodePtr & node_
...
-            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+        UInt64 pos;
+        if (constant_node->getValue().getType() == Field::Types::UInt64)
+        {
+            pos = constant_node->getValue().get<UInt64>();
+        }
+        else // Int64
+        {
+            auto value = constant_node->getValue().get<Int64>();
+            pos = value > 0 ? value : projection_nodes.size() + value + 1;
```
Comment (novikd) [inline src/Analyzer/Passes/QueryAnalysisPass.cpp:2172]: This line may lead to a overflow if `projection_nodes.size() + value + 1 < 0`

Comment (ucasfl) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp:2172]: I think it's ok, in such case, out of bound exception will be thrown later.

Comment (Avogar) [reply src/Analyzer/Passes/QueryAnalysisPass.cpp:2173]: Yes, the exception will be thrown anyway, but still let's add a separate check for possible overflow

**Review by Avogar — APPROVED** [review]

**Pull request merged**